state/presence/pruner_test.go

```diff
@@ -168,14 +168,21 @@ func waitForFirstChange(c *gc.C, watch <-chan Change, want Change) {
 // that the worker has stopped, and thus is no longer using its mgo
 // session before TearDownTest shuts down the connection.
 func assertStopped(c *gc.C, w worker.Worker) {
-	c.Assert(worker.Stop(w), jc.ErrorIsNil)
+	done := make(chan struct{})
+	go func() {
+		c.Check(worker.Stop(w), jc.ErrorIsNil)
+		close(done)
+	}()
+	select {
+	case <-done:
+		return
+	case <-time.After(testing.ShortWait):
+		c.Fatalf("failed to stop worker %v after %v", w, testing.ShortWait)
+	}
 }
 
 func (s *prunerSuite) TestDeepStressStaysSane(c *gc.C) {
 	FakePeriod(1)
-	// Each Pinger potentially grabs another socket to Mongo,
-	// which can exhaust connections. Somewhere around 5000 pingers on my
-	// machine everything starts failing because Mongo refuses new connections.
 	keys := make([]string, 500)
```

Comment:
So with the original 'direct' code, I could make it to 2000 (which should represent 4000 pings in some stage of alive/dead at any point in time) before thing started to fail. It also slows down dramatically (many minutes).
With the new code, I can bump it all the way up to 10,000 (20k active pingers), and the test completes in 37s without any problems.